### PR TITLE
cmd/snap-confine: add shutdown helper

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: debian/snapd.install [hunk 1]
```diff
@@ -2,3 +2,4 @@ usr/bin/snap
 usr/bin/snap-exec /usr/lib/snapd/
 usr/bin/snapctl
 usr/bin/snapd /usr/lib/snapd/
+usr/lib/snapd/system-shutdown
```
Comment (zyga) [inline debian/snapd.install]: I think this should be in the part that gets installed in core images (perhaps another package, perhaps one of the existing packages). At the very least the unit file should never be installed on regular systems.

Comment (chipaca) [inline debian/snapd.install]: I'll ask mvo about that. Note the unit file doesn't run on classic, though.

**Comment by chipaca** [comment]
Closing until I have time to make the requested changes.

**Comment by chipaca** [comment]
@zyga unfortunately this runs after everything else, including syslog, is off.

One way to test it is to boot a core with `kvm -stdio`, issue a `shutdown`, and grep the output. I don't think spread is up to that just yet?

File: debian/snapd.install (hunk 1 above)
Comment (chipaca) [inline debian/snapd.install]: mvo confirmed that thee is no 'part that gets intalled i core images', and that creating another package would be a pain for sru'ing and such

File: generate-packaging-dir [hunk 2]
```diff
@@ -11,20 +27,23 @@ ID="$1"
...
-rm -rf $DST
+rm -rf "$DST"
 mkdir -p "$DST"
 
-if ! git remote | grep upstream; then
-    git remote add upstream https://github.com/snapcore/snapd
+tree="$(tree1404)"
+if [ -z "$tree" ]; then
+    if ! git remote | grep upstream; then
+        git remote add upstream https://github.com/snapcore/snapd
```
Comment (pedronis) [inline generate-packaging-dir]: any particular reason to do this and a full fetch, instead of specifying all the bits in the fetch command?

Comment (chipaca) [inline generate-packaging-dir]: because I don't know how to do the latter?

Comment (pedronis) [inline generate-packaging-dir]: seems I commented in the less relevant spot

File: generate-packaging-dir [hunk 3]
```diff
@@ -1,6 +1,22 @@
...
+    if [ -z "${TRAVIS_PULL_REQUEST_SLUG:-}" ]; then
+        return 0
+    fi
+
+    local user="${TRAVIS_PULL_REQUEST_SLUG%/*}"
+
+    if ! git remote | grep -q "^$user$"; then
+        git remote add "$user" "https://github.com/$TRAVIS_PULL_REQUEST_SLUG" || return 0
+    fi
+    git fetch --quiet "$user" || return 0
```
Comment (pedronis) [inline generate-packaging-dir]: git fetch "https://github.com/$TRAVIS_PULL_REQUEST_SLUG" ${TRAVIS_PULL_REQUEST_BRANCH}-14.04

File: cmd/snap-confine/system-shutdown.c [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+#include <sys/reboot.h>		// reboot, RB_*
+#include <fcntl.h>		// open
+#include <errno.h>		// errno, sys_errlist
+
+#include "system-shutdown-utils.h"
+
+int main(int argc, char *argv[])
+{
+	errno = 0;
+	if (getpid() != 1) {
```
Comment (mvo5) [inline cmd/snap-confine/system-shutdown.c]: A small comment would be great, something like `systemd will *exec()* the shutdown binary` or something. When reading this I was thinking it means `getppid()`. Or maybe just moving the comment below that explains it all a bit up :)

File: debian/snapd.system-shutdown.service [hunk 5]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Ubuntu core (all-snaps) system shutdown helper setup service
+Before=umount.target
+DefaultDependencies=false
+# don't run on classic
+ConditionKernelCommandLine=snap_core
```
Comment (mvo5) [inline debian/snapd.system-shutdown.service]: \o/

**Pull request merged**